File: #0948.bag-of-tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {

        int size = tokens.size();
        if (size == 0) {
            return 0;
        }
   
        std::sort(tokens.begin(), tokens.end());
        if (tokens[0] > power) {
            return 0;
        }

        int result = 0;
        int index = size - 1;
        for (int i = 0; i <= index; i++) {
            power -= tokens[i];
            if (power >= 0 ) {
                result ++;
            } else {
                power += tokens[index--];
            }
        }
        return result;
       
    }
};

File: #0948.bag-of-tokens.js
/**
 * @param {number[]} tokens
 * @param {number} power
 * @return {number}
 */
var bagOfTokensScore = function(tokens, power) {
    tokens.sort(function(left, right) { return left - right });
    let left = 0;
    let right = tokens.length - 1;
    let score = 0;
    let max = 0;
    while (left <= right && power >= tokens[left]) {
        while (power >= tokens[left] && left <= right) {
            power -= tokens[left];
            score++;
            left++;
        }
        max = Math.max(max, score);
        if (score >= 1 && left < right) {
            power += tokens[right];
            score--;
            right--;
        }
    }
    return max;
};
